update_ip/services/cloudflare.py

```diff
@@ -0,0 +1,178 @@
+        :param dns_type: Domain type (A: for IPv4, AAAA: for IPv6...)
+        """
+        # Updates the known TLD names
+        tld.utils.update_tld_names()
+
+        # Cloudflare API URL
+        self.api_url = 'https://api.cloudflare.com/client/v4/'
+
+        # HTTP request and opener variables
+        self.request = None
+        self.opener = urllib2.build_opener(urllib2.HTTPHandler)
+
+        # Cloudflare data
+        self.email = email
+        self.api_key = api_key
+
+        # Domain Data
+        self.domain = None
+        self.dns_record = None
+        self.parse_domain(domain)
+
+        # DNS Type
+        self.record_type = dns_type
+
+        # Cloudflare internal data
+        self.zone_id = None
+        self.record_id = None
+
+    def update(self, domain, ip):
+        """Parse the given domain and extract the TLD
+        :param domain: Domain (or subdomain) to update
+        :param ip: IP address to update
+        """
+        self.parse_domain(domain)
+        self.update_ip(ip)
+
+    def find_domains(self, ip):
+        """Not implemented, since in cloudflare you can have multiple domains, subdomains, dns-records ... it would
+        not be safe to try to update them all
+        :param ip: IP address to update
+        """
+        raise NotImplementedError
+
+    def parse_domain(self, domain):
+        """Parse the given domain and extract the TLD
+        :param domain: Full domain
+        """
+        self.domain = tld.get_tld('http://' + domain)  # Convert from "any.sub.domain.tld" to "domain.tld"
+        self.dns_record = domain
+
+    def parse_result(self, data):
+        """Parse the JSON result and return a dict/list with data
+        :param data: JSON data to parse
+        """
+        return json.loads(data)
+
+    def update_ip(self, ip, second_try=False):
+        """Get the DNS Record info for selected sub-domain
+        :param ip: IP address to update
+        :param second_try: If Record_ID is empty try to get it, if everything fails in a second attempt, we throw an
+                            exception.
+        """
+        if self.record_id is None:
+            if second_try:
+                raise DNSServiceError(
+                    'Record ID is empty, we tried to find it, but it was not possible'
+                )
+            self.get_dns_record()
+            self.update_ip(ip, True)
+            return
+
+        self.create_request(self.generate_update_url(),
+                            json.dumps({'name': self.dns_record, 'type': self.record_type, 'content': ip})
+                            )
+
+        self.request.get_method = lambda: 'PUT'
+        result = self.opener.open(self.request)
+        pprint(result)
+        data = self.parse_result(result.read())
+        pprint(data)
+
+    def get_zones(self):
+        """Get the zone info for selected domain
+        """
+        self.create_request(self.generate_zones_list_url())
```

Comment:
I've created a new function __api_get_request that condense all this lines in a single function